### PR TITLE
fix: use vscode api to read processor groups files

### PR DESCRIPTION
**Pull request by ishche**
## Checklist:
- [x] Each of my commits contains one meaningful change
- [x] I have performed rebase of my branch on top of the development
- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

**Review by slavek-kucera — CHANGES_REQUESTED** [review]
Tests are failing on Windows.
```
Summary of all failing tests
 FAIL  src/__tests__/services/SettingsTest.spec.ts
  ● SettingsService evaluate variables › Evaluate fileDirname

    expect(received).toEqual(expected) // deep equality

    Expected: "a:\\toplevel/copybooks"
    Received: "/copybooks"

      113 |       "COBOL",
      114 |     );
    > 115 |     expect(paths[0]).toEqual(makefsPath("/toplevel") + "/copybooks");
          |                      ^
      116 |   });
      117 |
      118 |   test("Evaluate fileDirnameBasename", async () => {

      at Object.<anonymous> (src/__tests__/services/SettingsTest.spec.ts:115:22)

 FAIL  src/__tests__/services/ProcessorGroups.spec.ts
  ● Processor groups configuration provides lib path

    some issue with input param

      123 |   (globSync as any) = jest.fn().mockImplementation((config: string[]) => {
      124 |     if (config[0] === "/copy") return ["/copy-resolved-from-glob"];
    > 125 |     else throw Error("some issue with input param");
          |                ^
      126 |   });
      127 |   const result = await loadProcessorGroupCopybookPathsConfig(item, []);
      128 |   expect(result).toStrictEqual(["/copy-resolved-from-glob"]);

      at src/__tests__/services/ProcessorGroups.spec.ts:125:16
      at src/services/ProcessorGroups.ts:67:15
          at Array.map (<anonymous>)
      at map (src/services/ProcessorGroups.ts:66:6)
      at Object.<anonymous> (src/__tests__/services/ProcessorGroups.spec.ts:127:18)

  ● Processor groups configuration understend absolute paths

    some issue with input param

      136 |   (globSync as any) = jest.fn().mockImplementation((config: string[]) => {
      137 |     if (config[0] === "/abs") return ["/copy-resolved-from-glob"];
    > 138 |     else throw Error("some issue with input param");
          |                ^
      139 |   });
      140 |   const result = await loadProcessorGroupCopybookPathsConfig(item, []);
      141 |   expect(result).toStrictEqual(["/copy-resolved-from-glob"]);

      at src/__tests__/services/ProcessorGroups.spec.ts:138:16
      at src/services/ProcessorGroups.ts:67:15
          at Array.map (<anonymous>)
      at map (src/services/ProcessorGroups.ts:66:6)
      at Object.<anonymous> (src/__tests__/services/ProcessorGroups.spec.ts:140:18)

  ● Processor groups configuration provides copybook-extensions

    expect(received).toStrictEqual(expected) // deep equality

    - Expected  - 3
    + Received  + 1

    - Array [
    -   ".copy",
    - ]
    + Array []

      148 |   };
      149 |   const result = await loadProcessorGroupCopybookExtensionsConfig(item, []);
    > 150 |   expect(result).toStrictEqual([".copy"]);
          |                  ^
      151 | });
      152 |
      153 | it("Processor groups configuration provides copybook-file-encoding", async () => {

      at Object.<anonymous> (src/__tests__/services/ProcessorGroups.spec.ts:150:18)

  ● Processor groups configuration provides copybook-file-encoding

    expect(received).toStrictEqual(expected) // deep equality

    Expected: "UTF-8"
    Received: ""

      157 |   };
      158 |   const result = await loadProcessorGroupCopybookEncodingConfig(item, "");
    > 159 |   expect(result).toStrictEqual("UTF-8");
          |                  ^
      160 | });
      161 |
      162 | it("Processor groups configuration provides cobol-lsp.target-sql-backend", async () => {

      at Object.<anonymous> (src/__tests__/services/ProcessorGroups.spec.ts:159:18)

  ● Processor groups configuration provides cobol-lsp.target-sql-backend

    expect(received).toStrictEqual(expected) // deep equality

    Expected: "DATACOM_SERVER"
    Received: ""

      166 |   };
      167 |   const result = await loadProcessorGroupSqlBackendConfig(item, "");
    > 168 |   expect(result).toStrictEqual("DATACOM_SERVER");
          |                  ^
      169 | });
      170 |
      171 | it("Processor groups configuration provides dialect lib path", async () => {

      at Object.<anonymous> (src/__tests__/services/ProcessorGroups.spec.ts:168:18)

  ● Processor groups configuration provides dialect lib path

    expect(received).toStrictEqual(expected) // deep equality

    - Expected  - 3
    + Received  + 1

    - Array [
    -   "/daco",
    - ]
    + Array []

      174 |     "DaCo",
      175 |   );
    > 176 |   expect(result).toStrictEqual(["/daco"]);
          |                  ^
      177 | });
      178 |
      179 | it("Processor groups configuration matches program", async () => {

      at Object.<anonymous> (src/__tests__/services/ProcessorGroups.spec.ts:176:18)

  ● Processor groups configuration matches program

    expect(received).toStrictEqual(expected) // deep equality

    Expected: ["IDMS", "DaCo"]
    Received: {}

      183 |   };
      184 |   const result = await loadProcessorGroupDialectConfig(item, {});
    > 185 |   expect(result).toStrictEqual(["IDMS", "DaCo"]);
          |                  ^
      186 | });
      187 |
      188 | it("Processor groups configuration matches program relative to workspace", async () => {

      at Object.<anonymous> (src/__tests__/services/ProcessorGroups.spec.ts:185:18)

  ● Processor groups configuration matches program relative to workspace

    expect(received).toStrictEqual(expected) // deep equality

    - Expected  - 3
    + Received  + 1

    - Array [
    -   "IDMS",
    - ]
    + Array []

      192 |   };
      193 |   const result = await loadProcessorGroupDialectConfig(item, []);
    > 194 |   expect(result).toStrictEqual(["IDMS"]);
          |                  ^
      195 | });
      196 |
      197 | it("Processor groups configuration matches program with *", async () => {

      at Object.<anonymous> (src/__tests__/services/ProcessorGroups.spec.ts:194:18)

  ● Processor groups configuration provides compiler-options

    expect(received).toStrictEqual(expected) // deep equality

    Expected: ["QUALIFY(EXTEND)", "XMLPARSE(COMPAT)"]
    Received: ""

      219 |   };
      220 |   const result = await loadProcessorGroupCompileOptionsConfig(item, "");
    > 221 |   expect(result).toStrictEqual(["QUALIFY(EXTEND)", "XMLPARSE(COMPAT)"]);
          |                  ^
      222 | });
      223 |

      at Object.<anonymous> (src/__tests__/services/ProcessorGroups.spec.ts:221:18)


Test Suites: 2 failed, 31 passed, 33 total
Tests:       10 failed, 231 passed, 241 total
```

File: clients/cobol-lsp-vscode-extension/src/__tests__/commands/ClearCopybookCacheCommand.spec.ts [hunk 1]
```diff
@@ -35,6 +35,10 @@ jest.mock("vscode", () => ({
       delete: jest.fn().mockReturnValue(true),
       readDirectory: jest.fn().mockResolvedValue([["fileName", 2]]),
     },
+    getWorkspaceFolder: jest
+      .fn()
+      .mockReturnValue({ uri: { fsPath: "C:\\my\\workspace" } }),
+    workspaceFolders: [{ uri: { fsPath: "C:\\my\\workspace" } }],
```
Comment (slavek-kucera) [inline clients/cobol-lsp-vscode-extension/src/__tests__/commands/ClearCopybookCacheCommand.spec.ts]: Why is this needed? The `clearCache` function does not depend on the `workspaceFolders`.

File: clients/cobol-lsp-vscode-extension/src/__tests__/services/LanguageClientServiceTest.spec.ts [hunk 2]
```diff
@@ -34,6 +34,10 @@ jest.mock("vscode", () => ({
     getConfiguration: jest.fn().mockReturnValue({
       get: jest.fn().mockReturnValue(0),
     }),
+    getWorkspaceFolder: jest
+      .fn()
+      .mockReturnValue({ uri: { fsPath: "C:\\my\\workspace" } }),
+    workspaceFolders: [{ uri: { fsPath: "C:\\my\\workspace" } }],
```
Comment (slavek-kucera) [inline clients/cobol-lsp-vscode-extension/src/__tests__/services/LanguageClientServiceTest.spec.ts]: not needed?

File: clients/cobol-lsp-vscode-extension/src/__tests__/services/BridgeForGitSupport.spec.ts [hunk 3]
```diff
@@ -35,6 +35,9 @@ jest.mock("vscode", () => ({
     getConfiguration: jest.fn().mockReturnValue({
       get: jest.fn(),
     }),
+    getWorkspaceFolder: jest
+      .fn()
+      .mockReturnValue({ uri: { fsPath: "C:\\my\\workspace" } }),
```
Comment (slavek-kucera) [inline clients/cobol-lsp-vscode-extension/src/__tests__/services/BridgeForGitSupport.spec.ts]: not needed

File: clients/cobol-lsp-vscode-extension/src/__tests__/services/util/ConfigurationWatcher.spec.ts [hunk 4]
```diff
@@ -29,6 +29,11 @@ jest.mock("vscode", () => ({
       get: jest.fn().mockReturnValueOnce(undefined).mockReturnValue("JAVA"),
       update: jest.fn(),
     }),
+    getWorkspaceFolder: jest
+      .fn()
+      .mockReturnValue({ uri: { fsPath: "C:\\my\\workspace" } }),
+    workspaceFolders: [{ uri: { fsPath: "C:\\my\\workspace" } }],
+
```
Comment (slavek-kucera) [inline clients/cobol-lsp-vscode-extension/src/__tests__/services/util/ConfigurationWatcher.spec.ts]: this can be removed as well.

File: clients/cobol-lsp-vscode-extension/src/services/ProcessorGroupsLoader.ts [hunk 5]
```diff
@@ -65,56 +64,65 @@ const ProcessorGroupModel = t.intersection([
...
-  const cfgPath = path.join(ws[0], PROCESSOR_GROUP_FOLDER);
-  const pgmCfgPath = path.join(cfgPath, PROCESSOR_GROUP_PGM);
-  if (!fs.existsSync(pgmCfgPath)) {
+
+  const ws = workspace.getWorkspaceFolder(documentUri);
+  if (ws === undefined) {
     return EMPTY;
   }
+  const cfgPath = Uri.joinPath(ws.uri, PROCESSOR_GROUP_FOLDER);
+  const pgmCfgPath = Uri.joinPath(cfgPath, PROCESSOR_GROUP_PGM);
```
Comment (slavek-kucera) [inline clients/cobol-lsp-vscode-extension/src/services/ProcessorGroupsLoader.ts]: `joinPath` can actually join multiple segments... but I guess it will break all those mocks, right?

File: clients/cobol-lsp-vscode-extension/src/services/copybook/CopybookDownloadService.ts [hunk 6]
```diff
@@ -140,10 +140,12 @@ export class CopybookDownloadService {
...
-      return await this.e4eDownloader?.getE4ECopyBookLocation(
-        copybookName,
-        documentUri,
-      );
+      return (
+        await this.e4eDownloader?.getE4ECopyBookLocation(
+          copybookName,
+          documentUri,
+        )
+      )?.toString();
```
Comment (slavek-kucera) [inline clients/cobol-lsp-vscode-extension/src/services/copybook/CopybookDownloadService.ts]: I suggest either
```suggestion
      const copybookUri = await this.e4eDownloader?.getE4ECopyBookLocation(
        copybookName,
        documentUri,
      );
      return copybookUri?.toString();
```
or
```suggestion
      return this.e4eDownloader?.getE4ECopyBookLocation(
        copybookName,
        documentUri,
      ).then(uri => uri?.toString());
```

File: clients/cobol-lsp-vscode-extension/src/__tests__/services/copybook/CopybookResolveURITest.spec.ts [hunk 7]
```diff
@@ -32,63 +31,27 @@ const CPY_FOLDER_NAME = ".cobcopy";
...
-      readFile: jest.fn().mockImplementation(() => {
-        throw { code: "FileNotFound" };
+jest.mock("vscode", () => {
+  return {
+    Uri: require("../../../__mocks__/UriMock").Uri,
+    window: {
+      createOutputChannel: jest.fn().mockReturnValue({
+        appendLine: jest.fn(),
       }),
     },
```
Comment (slavek-kucera) [inline clients/cobol-lsp-vscode-extension/src/__tests__/services/copybook/CopybookResolveURITest.spec.ts]: not needed

File: clients/cobol-lsp-vscode-extension/src/__tests__/services/copybook/CopybookResolveURITest.spec.ts [hunk 8]
```diff
@@ -32,63 +31,27 @@ const CPY_FOLDER_NAME = ".cobcopy";
...
+    workspace: {
+      fs: {
+        readFile: jest.fn().mockImplementation(() => {
+          throw { code: "FileNotFound" };
+        }),
+      },
+      getWorkspaceFolder: jest
+        .fn()
+        .mockReturnValue({ uri: { fsPath: "C:\\my\\workspace" } }),
+      workspaceFolders: [{ uri: { fsPath: "C:\\my\\workspace" } }],
```
Comment (slavek-kucera) [inline clients/cobol-lsp-vscode-extension/src/__tests__/services/copybook/CopybookResolveURITest.spec.ts]: not needed

File: clients/cobol-lsp-vscode-extension/src/__tests__/services/ProcessorGroups.spec.ts [hunk 9]
```diff
@@ -25,102 +26,80 @@ import { globSync } from "glob";
...
+beforeAll(() =>
+  Object.defineProperty(process, "platform", {
+    value: "darwin",
+  }),
+);
+afterAll(() =>
+  Object.defineProperty(process, "platform", {
+    value: platform,
+  }),
+);
```
Comment (slavek-kucera) [inline clients/cobol-lsp-vscode-extension/src/__tests__/services/ProcessorGroups.spec.ts]: What part of the code needs this?

Comment (ishche) [reply clients/cobol-lsp-vscode-extension/src/__tests__/services/ProcessorGroups.spec.ts]: UriMock as Uri handles paths->fsPath conversion differently. At least I understand it so https://github.com/microsoft/vscode-uri/blob/6dec22d7dcc6c63c30343d3a8d56050d0078cb6a/src/uri.ts#L564

So to pass tests on windows machine, im set platform to non windows.

Comment (ishche) [reply clients/cobol-lsp-vscode-extension/src/__tests__/services/ProcessorGroups.spec.ts]: I guess I can avoid it by checking for column

Comment (slavek-kucera) [reply clients/cobol-lsp-vscode-extension/src/__tests__/services/ProcessorGroups.spec.ts]: I think that the root cause of this whole problem is that we have 2 versions of the test - one for windows, one for the rest, but we are always running both.
The only difference between those tests are 2 absolute paths. I suggest merging the two files into one.

File: clients/cobol-lsp-vscode-extension/src/__tests__/services/BridgeForGitSupport.spec.ts (hunk 3 above)
Comment (ishche) [reply clients/cobol-lsp-vscode-extension/src/__tests__/services/BridgeForGitSupport.spec.ts]: Why do you think so?

Comment (slavek-kucera) [reply clients/cobol-lsp-vscode-extension/src/__tests__/services/BridgeForGitSupport.spec.ts]: Because it was not used by `loadProcessorsConfigForDocument` in the commit the comment was made on.

**Review by slavek-kucera — APPROVED** [review]
